### PR TITLE
Constructors: fixes to maintain needed private fields.

### LINKED ISSUE #772: Constructors removes private fields used only in inner classes
For example `ctx` field here: 
https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/typer/VarianceChecker.scala#L21

I'm preparing a fix.

### PR DESCRIPTION
**Pull request by DarkDimius**
fixes #772

### CODE REVIEW

**Comment by DarkDimius** [comment]
> [error] (*:update) sbt.ResolveException: unresolved dependency: org.scala-sbt#relation;0.13.5: not found

rebuild

**Comment by DarkDimius** [comment]
@odersky please review

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 1]
```diff
@@ -33,6 +33,61 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+        tree match {
+          case Ident(_) | Select(This(_), _) =>
+            def inConstructor = {
+              val method = ctx.owner.enclosingMethod
+              method.isPrimaryConstructor && ctx.owner.enclosingClass == owner
+            }
+            if (inConstructor && (sym.is(ParamAccessor) || seenPrivateVals.contains(sym))) {
+              // used inside constructor, accessed on this,
+              // could use constructor argument instead, no need to retain field
+              println("hoha")
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/Constructors.scala]: hoha! You probably don't want to keep this :)

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 2]
```diff
@@ -33,6 +33,61 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+  // Collect all private parameter accessors and value definitions that need
+  // to be retained. There are several reasons why a parameter accessor or
+  // definition might need to be retained:
+  // 1. It is accessed after the constructor has finished
+  // 2. It is accessed before it is defined
+  // 3. It is accessed on an object other than `this`
+  // 4. It is a mutable parameter accessor
+  // 5. It is has a wildcard initializer `_`
+  private var retainedPrivateVals = mutable.Set[Symbol]()
+  private var seenPrivateVals = mutable.Set[Symbol]()
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/Constructors.scala]: `retainedPrivateVals` and `seenPrivateVals` do not need to be `var`.

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 3]
```diff
@@ -33,6 +33,58 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+  private def markUsedPrivateSymbols(tree: RefTree)(implicit ctx: Context): Unit = {
+
+    val sym = tree.symbol
+    def retain =
+      retainedPrivateVals.add(sym)
+
+    if (mightBeDropped(sym) && sym.owner.isClass) {
+      val owner = sym.owner.asClass
+
+        tree match {
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/Constructors.scala]: The return value of this match is not used anywhere.

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 4]
```diff
@@ -33,6 +33,58 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+          case Ident(_) | Select(This(_), _) =>
+            def inConstructor = {
+              val method = ctx.owner.enclosingMethod
+              method.isPrimaryConstructor && ctx.owner.enclosingClass == owner
+            }
+            if (inConstructor && (sym.is(ParamAccessor) || seenPrivateVals.contains(sym))) {
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/Constructors.scala]: This block is empty.

File: src/dotty/tools/dotc/transform/Constructors.scala (hunk 3 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/Constructors.scala]: this match returns Unit. How could it be used?

File: src/dotty/tools/dotc/transform/Constructors.scala (hunk 4 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/Constructors.scala]: This is intended this block contains comment on why nothing happens here

File: src/dotty/tools/dotc/transform/Constructors.scala (hunk 3 above)
Comment (smarter) [inline src/dotty/tools/dotc/transform/Constructors.scala]: Ah ok, I missed that `retain` had side-effects, I suggest replacing `def retain` by `def retain()` and using `retain()` to follow the scala style.

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 5]
```diff
@@ -33,6 +33,58 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+  private val retainedPrivateVals = mutable.Set[Symbol]()
+  private val seenPrivateVals = mutable.Set[Symbol]()
+
+  private def markUsedPrivateSymbols(tree: RefTree)(implicit ctx: Context): Unit = {
+
+    val sym = tree.symbol
+    def retain =
+      retainedPrivateVals.add(sym)
+
+    if (mightBeDropped(sym) && sym.owner.isClass) {
```
Comment (odersky) [inline src/dotty/tools/dotc/transform/Constructors.scala]: I'd reverse the order of the two conditions in the &&. mightBeDropped makes sense only for class members.

File: src/dotty/tools/dotc/transform/Constructors.scala [hunk 6]
```diff
@@ -33,6 +33,58 @@ class Constructors extends MiniPhaseTransform with SymTransformer { thisTransfor
...
+  }
+
+  override def transformSelect(tree: tpd.Select)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
+    markUsedPrivateSymbols(tree)
+    tree
+  }
+
+  override def transformValDef(tree: tpd.ValDef)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
+    if (mightBeDropped(tree.symbol))
+      (if (isWildcardStarArg(tree.rhs)) retainedPrivateVals else seenPrivateVals) += tree.symbol
```
Comment (odersky) [inline src/dotty/tools/dotc/transform/Constructors.scala]: Why the `isWildcardStarArg` test?

**Comment by odersky** [comment]
Otherwise LGTM

**Pull request merged**